src/passes/MultiMemoryLowering.cpp

```diff
@@ -118,35 +118,76 @@ struct MultiMemoryLowering : public Pass {
+        curr->ptr = builder.makeLocalGet(ptrIdx, parent.pointerType);
+        Expression* boundsCheck = builder.makeIf(
+          builder.makeBinary(
+            Abstract::getBinary(parent.pointerType, Abstract::GtU),
+            builder.makeBinary(
+              // ptr + offset (ea from wasm spec) + bit width
+              // two builder Adds, we'll add the first two operands in the first
+              // add and then add the third operand in the second add
+              Abstract::getBinary(parent.pointerType, Abstract::Add),
+              builder.makeBinary(
+                Abstract::getBinary(parent.pointerType, Abstract::Add),
+                builder.makeLocalGet(ptrIdx, parent.pointerType),
+                builder.makeConstPtr(curr->offset, parent.pointerType)),
+              builder.makeConstPtr(curr->bytes, parent.pointerType)),
+            builder.makeCall(
+              parent.memorySizeNames[idx], {}, parent.pointerType)),
+          builder.makeUnreachable());
+        replacement = builder.makeBlock({ptrSet, boundsCheck, curr});
+        replaceCurrent(replacement);
       }
 
-      // We diverge from the spec here and are not trapping if the offset + type
-      // / 8 is larger than the length of the memory's data. Warning,
-      // out-of-bounds loads and stores can read junk out of or corrupt other
-      // memories instead of trapping
       void visitStore(Store* curr) {
         auto idx = parent.memoryIdxMap.at(curr->memory);
         auto global = parent.getOffsetGlobal(idx);
         curr->memory = parent.combinedMemory;
         if (!global) {
           return;
         }
-        curr->ptr = builder.makeBinary(
-          Abstract::getBinary(parent.pointerType, Abstract::Add),
-          builder.makeGlobalGet(global, parent.pointerType),
-          curr->ptr);
+        Expression* replacement;
+        Index ptrIdx = Builder::addVar(getFunction(), parent.pointerType);
+        Expression* ptrSet = builder.makeLocalSet(
+          ptrIdx,
+          builder.makeBinary(
+            Abstract::getBinary(parent.pointerType, Abstract::Add),
+            builder.makeGlobalGet(global, parent.pointerType),
+            curr->ptr));
+        curr->ptr = builder.makeLocalGet(ptrIdx, parent.pointerType);
+        Expression* boundsCheck = builder.makeIf(
+          builder.makeBinary(
+            Abstract::getBinary(parent.pointerType, Abstract::GtU),
+            builder.makeBinary(
+              // ptr + offset (ea from wasm spec) + bit width
+              // two builder Adds, we'll add the first two operands in the first
+              // add and then add the third operand in the second add
+              Abstract::getBinary(parent.pointerType, Abstract::Add),
+              builder.makeBinary(
+                Abstract::getBinary(parent.pointerType, Abstract::Add),
+                builder.makeLocalGet(ptrIdx, parent.pointerType),
+                builder.makeConstPtr(curr->offset, parent.pointerType)),
```

Comment:
Probably worth a comment that the offset computation is not precise according to the spec. In the spec offsets do not overflow as twos-complement, but `i32.add` does. Concretely, a load from address `1000` with offset `0xffffffff` should actually trap, as the combined number is greater than 32 bits. But with an add, `1000 + 0xffffffff = 999` due to overflow, which would not trap.

In theory we could compute this like the spec, by expanding the i32s to i64s and adding there (where we won't overflow). But I think for now a comment is enough, since it's not that important to keep trapping programs trapping (as opposed to working programs working).